### ISSUE TITLE
gslib/storage_url.py references text_util module without importing it

### ISSUE THREAD

**Issue by houglum** [GoogleCloudPlatform/gsutil#530]
When refactoring gslib/util.py into smaller components, it looks like the method `GenerationFromUrlAndString` got moved to gslib/storage_url.py and was changed to use a method in the `text_util` module... but that module never got imported.  This is only shows up when operating on S3 resources.  The resulting error is:

`global name 'text_util' is not defined`

This affects v4.31 of gsutil.

**Comment by mfschwartz** [comment]
The fix is implemented, and will be included in gsutil v4.32.

**Comment by HarmeetSokhi** [comment]
Thanks @mfschwartz . Do you know when is the release pending?

**Comment by nestoru** [comment]
In the meantime while this is resolved you can patch it yourself. Assuming you have a default installation of gcloud you can make a copy of  ~/google-cloud-sdk/platform/gsutil/gslib/storage_url.py and add the corresponding line that addresses the missing dependency:
```
$ diff -c ~/storage_url.py.old  ~/google-cloud-sdk/platform/gsutil/gslib/storage_url.py
*** ~/storage_url.py.old     2018-05-31 19:13:45.718235433 +0000
--- ~/google-cloud-sdk/platform/gsutil/gslib/storage_url.py  2018-05-30 21:23:15.895535824 +0000
***************
*** 21,26 ****
--- 21,27 ----
  import stat
  
  from gslib.exception import InvalidUrlError
+ from gslib.utils import text_util
  
  # Matches provider strings of the form 'gs://'
  PROVIDER_REGEX = re.compile(r'(?P<provider>[^:]*)://$')
```

**Comment by mfschwartz** [comment]
You can patch it yourself but just to make sure it's clear: The GCS team doesn't support gsutil in a patched state (or running gsutil directly from a GitHub enlistment). If you have problems you'll need to revert to one of the distributed versions of gsutil.

**Comment by HarmeetSokhi** [comment]
@nestoru 
now we get this global name 'PleaseRetryException' is not defined

**Comment by houglum** [comment]
v4.32 went out on Friday - we missed the deadline to get bundled with this week's version of gcloud, so v4.32 won't be available as part of gcloud until next week.  It's available for standalone installations now, though.

Also, the PleaseRetryException is another bug, with the same general root cause (method being referenced with no corresponding import). This one happens because of a timeout from S3, so we couldn't easily test for it.  I'm puzzled as to why the linter didn't pick up this issue when that line was submitted... could be that I forgot to run it when adding that patch in? Either way, I'll file a separate issue for that.

**Comment by nestoru** [comment]
@houglum I suggest adding linter to the release script.
@mfschwartz has been working for me probably because I am not using the 'PleaseRetryException'? This is the command that works for me:
`gsutil -m rsync -C -r -d gs://src s3://tgt`

**Comment by houglum** [comment]
Agreed. We currently have it documented as a step of our review/push process, but it's easy for a human to forget a step :/

Also, you're unlikely to see the "PleaseRetryException" error unless you get a request timeout from S3.

**Comment by HarmeetSokhi** [comment]
Thanks @houglum . @nestoru  Let me know what to add to gsutil for this issue.  We  have not been able sync files from 18th May.  Thanks  everyone.

**Comment by nestoru** [comment]
@HarmeetSokhi what I explained above should work if you do not get an s3 timeout. If you can afford synching whole paths instead of the whole root s3 root you will have more chances to succeed because there will be fewer fils to sync. If it fails with timeout of course you can manually retry. We all hope the final gcloud update will be out soon (next week per @houglum). Perhaps @houglum can point to instructions to perform a standalone installation?

**Comment by houglum** [comment]
Those instructions are here: https://cloud.google.com/storage/docs/gsutil_install#alt-install

But please note that using both installation methods often results in confusion, specifically around auth and credentials (see #457).  I honestly don't recommend having two different installation styles (gcloud and non-gcloud) on the same machine, for the reasons mentioned there.

Also, in reference to my comment above (https://github.com/GoogleCloudPlatform/gsutil/issues/530#issuecomment-394600677), I fixed issue #533, but that fix won't be present until v4.33 is released (hoping to do so soon, but we're currently blocked on release test issues).

**Comment by HarmeetSokhi** [comment]
Hi, Is the fix released?

**Comment by houglum** [comment]
Unfortunately, no - still seeing hiccups with our release tests.  I'm working with the gcloud team to get our testing issues figured out this week so that we can do another release.

**Comment by houglum** [comment]
The v4.33 gsutil release is out for standalone gsutil, but I believe we missed the cut for the coming week's (week of 06/25) gcloud release.  There won't be a gcloud release on the following week due to the July 4th holiday, but gsutil v4.33 should be included in the next gcloud release (v208, I believe).
